Add env selector button, and new kafka metrics

gburges opened the pull request.

gburges committed:
Add env selector button, and new kafka metrics

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -85,6 +108,24 @@ export function KafkaInfoComponent() {
+    <FormControl>
+      <InputLabel id="cluster-select-label">Cluster</InputLabel>
+      <Select labelId="cluster-select-label" id="cluster-select" value={currentCluster} onChange={handleClusterChange}>
+        {Object.keys(clusterMap).map((key) => (
```
Because you are iterating over the clusters it might make sense to have an array of clusters rather than an object.

adamrdrew replied:
Also, you could store the cluster object in state instead of an object name for looking up in the in the cluster object.

gburges replied:
is this how you meant?

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -37,10 +56,10 @@ export function KafkaInfoComponent() {
     setLoading(true);
     // Directly query a prometheus endpoint for metric data
-    fetch(`${backendUrl}/api/proxy/kafka-lag/query?query=aws_kafka_max_offset_lag_sum`)
+    fetch(`${backendUrl}/api/proxy/${proxy}/query?query=aws_kafka_sum_offset_lag_sum`)
```
It would be good to make this authenticated while you are in here.

adamrdrew reviewed (approved):
I'd approve it as is, but there are a few opportunities for improvement if you'd like to do them.

wcmitchell commented:
Working on testing this locally, but overall looks good to me. @adamrdrew, thoughts?

wcmitchell commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -17,6 +22,21 @@ import { useEntity } from '@backstage/plugin-catalog-react';
 import { useApi, configApiRef } from '@backstage/core-plugin-api';
 
+const clusterMap = [
+  stage: {
```
I'm getting syntax failures here, not sure if it's my dev env or something valid. It doesn't seem to be liking named objects within the array?

Working through a couple of other errors in my env, but this seems to be the major break for me. @adamrdrew seeing similar?

adamrdrew commented on app-config.yaml:
```diff
@@ -51,8 +51,10 @@ proxy:
   #     changeOrigin: true
   endpoints:
-    '/kafka-lag':
+    '/kafka-lag-stage':
```
I don't love this naming scheme. It feels like it should be 

`
'kafka-lag/stage`:
     ...
`kafka-lag/prod`:
    ....
```

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -17,6 +22,21 @@ import { useEntity } from '@backstage/plugin-catalog-react';
+  stage: {
+    proxy: `kafka-lag-stage`,
+    name: 'Stage',
+    token: '${STAGE_PROMETHEUS_TOKEN}',
```
This still looks wrong. It looks like you are either trying to do string interpolation incorrectly, or attempting an incorrect way of doing env var interpolation.

This also just doesn't make sense because you can put the token in the proxy by adding an auth header.

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -37,10 +58,15 @@ export function KafkaInfoComponent() {
     setLoading(true);
     // Directly query a prometheus endpoint for metric data
-    fetch(`${backendUrl}/api/proxy/kafka-lag/query?query=aws_kafka_max_offset_lag_sum`)
+    fetch(`${backendUrl}/api/proxy/${proxy}/query?query=aws_kafka_sum_offset_lag_sum`, {
```
Please use backstage authenticated fetchApi

See https://github.com/RedHatInsights/backstage-plugin-visual-qontract/blob/b82c3f02659b85f9524ca717782ee415f0282403/plugins/visual-qontract/src/components/SLOComponent/SLOGauge.tsx#L15

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -37,10 +58,15 @@ export function KafkaInfoComponent() {
-    fetch(`${backendUrl}/api/proxy/kafka-lag/query?query=aws_kafka_max_offset_lag_sum`)
+    fetch(`${backendUrl}/api/proxy/${proxy}/query?query=aws_kafka_sum_offset_lag_sum`, {
+      method: "GET", 
+      headers: {"Authorization": `Bearer ${token}`, 
```
As per my comment about the tokens - this should be handled on the back end.

See https://github.com/RedHatInsights/backstage-plugin-visual-qontract/blob/b82c3f02659b85f9524ca717782ee415f0282403/app-config.yaml#L61

adamrdrew reviewed (changes requested):
There's a decent amount here that needs to be iterated on before we can merge.

wcmitchell committed:
Env switch overhaul

Add dynamic config, fit env selector into existing UI, make lag metric
configurable

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

wcmitchell committed:
Add sample component

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

adamrdrew commented on app-config.yaml:
```diff
@@ -43,6 +43,16 @@ integrations:
     #   apiBaseUrl: https://ghe.example.net/api/v3
     #   token: ${GHE_TOKEN}
 
```
You are going to need to add

```
dangerouslyAllowUnverifiedAuth: true
```

To the top level app-config.yaml. This is because backstage disabled the ability to interact with backstage if you aren't logged in as a user with a catalog entity. Adding this in the dev repo allows that.

adamrdrew commented on app-config.yaml:
```diff
@@ -51,8 +61,14 @@ proxy:
   endpoints:
-    '/kafka-lag':
+    '/kafka-lag/stage':
       target: 'https://prometheus-server/api/v1/'
```
These URLs should be env vars.

adamrdrew commented on app-config.yaml:
```diff
@@ -43,6 +43,16 @@ integrations:
 
+kafkaInfo:
+  # Metric we'll use for tracking topic use and lag
+  lagMetric: 'kafka_consumergroup_group_topic_sum_lag'
```
Should this be an env var? If someone in the community or in a different BU or a different Kafka provider used this plugin would it work with this hard coded?

wcmitchell replied:
Ah yeah, that's a good idea. I just figured they'd update the app-config, but env var makes more sense

adamrdrew commented on catalog-info.yaml:
```diff
@@ -1,13 +1,37 @@
+
+  annotations:
+    # argocd/app-name: 'janus-idp'
+    kafka-info/consumer-groups: zomg
```
This is good, but please add a comment to the README that mentions changing this to whatever a user would need.

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -8,15 +8,91 @@ import {
+  ListItemText,
+  Menu,
+  IconButton,
+  ListItemSecondaryAction,
```
This is imported but never used.

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -27,49 +103,72 @@ export function KafkaInfoComponent() {
+    const fetchLags = (proxy: string) => {
+      setLoading(true);
+      // Directly query a prometheus endpoint for metric data
+      fetch(`${backendUrl}/api/proxy/${proxy}/api/v1/query?query=${lagMetric}`, {
```
We need to switch from using vanilla `fetch` to the Backstage provided `fetchApi.fetch` instead so we can make use of authenticated routes.

See here https://github.com/RedHatInsights/backstage-plugin-visual-qontract/blob/master/plugins/webrca-frontend/src/components/WebRCAFetchComponent/WebRCAFetchComponent.tsx#L104 for an example.

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -111,7 +209,12 @@ export function KafkaInfoComponent() {
 
   return (
-    <InfoCard title={title} noPadding>
+    <InfoCard title={title}>
```
We talked about unifying these `<InfoCard/>` components so that you don't need to recreate it multiple times. That's still a viable option. I wouldn't die on this hill, it isn't a ton of lines being duplicated, but it is some, and there's some differences, see the `<Grid/>` components in `if (loading)` - I suggest taking a pass at it.

adamrdrew reviewed (changes requested):
Definitely moving in the right direction but there are multiple changes required here.

Additionally, we are running on an old verison of backstage here: 1.28.0. Our goal is to target the same version of backstage that our version of RHDH is running, which is 1.32.6 so you are going to need to upgrade the backstage bits.

The good news it is quite easy really https://backstage.io/docs/getting-started/keeping-backstage-updated/

Basically:
```
npx backstage-cli versions:bump 1.32.6
```

You may encounter some problems:

* It may want a different version of node than you are using. I suggest using nvm to manage your node versions
* It may bomb out on start with messages about imports with the word `alpha` on the end of the name like `backend.add(import('@backstage/plugin-permission-backend/alpha'));` The good news is it is easy to fix: find those imports and remove `/alpha`. Some backstage stuff that was previously alpha is now GA.

If you run into other weirdness you can also delete `node_modules` and `plugins/kafka-info/node_modules`

That's it! Keep going and I look forward to the next revision.

wcmitchell committed:
Use backstage fetchApi

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

wcmitchell committed:
Clarify prometheus endpoint in sample config

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

wcmitchell committed:
Further parameterize dev app-config

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

wcmitchell committed:
Overhaul everything!

Ok, so, this is a huge refactor, but overall works a lot more cleanly.

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

adamrdrew commented on plugins/kafka-info/src/components/KafkaInfoComponent/KafkaInfoComponent.tsx:
```diff
@@ -1,118 +1,198 @@
-    setFilteredResponse(filteredGroup);
-    setLoading(false);
-  }, [metricResponse]);
+  const { value, loading, error } = useAsync(async (): Promise<
```
this isn't wrong, but its odd. We do fetches in normal side effects or functions called by side effects all the time. I'm curious why this is implemented like this. It's OK, maybe even good because you don't have to manage loading state manually, but why did you reach for this.

Also, this fetch needs more robust error handling. Right now if you get something you don't expect back from the server, like HTML rather than json the attempt to parse the JSON bombs out and crashes the web app in the client browser.

adamrdrew reviewed (approved):
No matter what I do I can't run this locally. But, it looks sane to me.
